x-pack/plugins/lens/public/xy_visualization/get_legend_action.tsx

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React from 'react';
+import type { LegendAction, XYChartSeriesIdentifier } from '@elastic/charts';
+import type { LayerArgs } from './types';
+import type { LensMultiTable, LensFilterEvent, FormatFactory } from '../types';
+import { LegendActionPopover } from '../shared_components';
+
+export const getLegendAction = (
+  filteredLayers: LayerArgs[],
+  tables: LensMultiTable['tables'],
+  onFilter: (data: LensFilterEvent['data']) => void,
+  formatFactory: FormatFactory,
+  layersAlreadyFormatted: Record<string, boolean>
+): LegendAction =>
+  React.memo(({ series: [xySeries] }) => {
+    const series = xySeries as XYChartSeriesIdentifier;
+    const layer = filteredLayers.find((l) =>
+      series.seriesKeys.some((key: string | number) => l.accessors.includes(key.toString()))
+    );
+
+    if (!layer || !layer.splitAccessor) {
+      return null;
+    }
+
+    const splitLabel = series.seriesKeys[0] as string;
+    const accessor = layer.splitAccessor;
+
+    const table = tables[layer.layerId];
+    const splitColumn = table.columns.find(({ id }) => id === layer.splitAccessor);
+    const formatter = formatFactory(splitColumn && splitColumn.meta?.params);
+
+    const currentSplitFormatter =
+      layersAlreadyFormatted[accessor] && splitColumn
+        ? formatFactory(splitColumn.meta.params)
+        : formatter;
+
+    const rowIndex = table.rows.findIndex((row) => {
+      if (layersAlreadyFormatted[accessor]) {
+        // stringify the value to compare with the chart value
+        return currentSplitFormatter.convert(row[accessor]) === splitLabel;
+      }
+      return row[accessor] === splitLabel;
+    });
+
+    if (rowIndex < 0) return null;
+
+    const data = [
+      {
+        row: rowIndex,
+        column: table.columns.findIndex((col) => col.id === accessor),
+        value: accessor ? table.rows[rowIndex][accessor] : splitLabel,
+        table,
+      },
+    ];
+
+    const context: LensFilterEvent['data'] = {
+      data,
+    };
+
+    return (
+      <LegendActionPopover
+        label={formatter ? formatter.convert(splitLabel) : splitLabel}
```

Comment:
We need to check `layersAlreadyFormatted` here as well - otherwise it will format twice and break
<img width="381" alt="Screenshot 2021-06-21 at 11 34 24" src="https://user-images.githubusercontent.com/1508364/122741234-0a496e80-d285-11eb-985d-915a6f585bf3.png">

Reply:
Thanx Joe, and for merging with upstream :)